chisel/src/main.rs

```diff
@@ -1,30 +1,344 @@
+    process::exit(-1);
+}
+
+fn yaml_configure(yaml: &String) -> Result<ChiselContext, &'static str> {
+    if let Ok(ruleset) = serde_yaml::from_str::<Value>(yaml.as_str()) {
+        ChiselContext::from_ruleset(&ruleset)
+    } else {
+        Err(ERR_FAILED_PARSE_CONFIG)
+    }
+}
+
+/// Helper that tries both translation methods in the case that a module cannot implement one of
+/// them.
+fn translate_module<T>(module: &mut Module, translator: T) -> Result<bool, &'static str>
+where
+    T: ModuleTranslator,
+{
+    // NOTE: The module must return an Err (in the case of failure) without mutating the module or nasty stuff happens.
+    if let Ok(ret) = translator.translate_inplace(module) {
+        Ok(ret)
+    } else if let Ok(new_module) = translator.translate(module) {
+        if new_module != *module {
+            // hacky expensive compare
+            *module = new_module;
+            Ok(true) // Needs a better way of passing whether the module was translated or not.
+        } else {
+            Ok(false)
+        }
+    } else {
+        Err("Module translation failed")
+    }
+}
+
+fn execute_module(context: &ModuleContext, module: &mut Module) -> bool {
+    let (conf_name, conf_preset) = context.fields();
+    let preset = conf_preset.clone();
+
+    let mut is_translator = false; // Flag representing if the module is a translator
+    let name = conf_name.as_str();
+    let ret = match name {
+        "verifyexports" => {
+            if let Ok(chisel) = VerifyExports::with_preset(&preset) {
+                Ok(chisel.validate(module).unwrap_or(false))
+            } else {
+                Err("Invalid preset")
+            }
+        }
+        "verifyimports" => {
+            if let Ok(chisel) = VerifyImports::with_preset(&preset) {
+                Ok(chisel.validate(module).unwrap_or(false))
+            } else {
+                Err("Invalid preset")
+            }
+        }
+        "checkstartfunc" => {
+            //NOTE: checkstartfunc takes a bool for configuration. false by default for now.
+            let chisel = CheckStartFunc::new(false);
+            let ret = chisel.validate(module).unwrap_or(false);
+            Ok(ret)
+        }
+        "trimexports" => {
+            is_translator = true;
+            if let Ok(chisel) = TrimExports::with_preset(&preset) {
+                translate_module(module, chisel)
+            } else {
+                Err("Invalid preset")
+            }
+        }
+        "remapimports" => {
+            is_translator = true;
+            if let Ok(chisel) = RemapImports::with_preset(&preset) {
+                translate_module(module, chisel)
+            } else {
+                Err("Invalid preset")
+            }
+        }
+        "deployer" => {
+            is_translator = true;
+            let mut payload = Vec::new();
+            module.clone().serialize(&mut payload).unwrap(); // TODO: handle serialization failure
```

Comment:
Apparently `module` is deserialised from the input binary so it is very unlikely that serialising that would fail.

Reply:
ok, should i remove the comment?

Reply:
Maybe just mention this fact. "This shouldn't fail, but need to dobule check"